In [1]:
import numpy as np
import matplotlib.pyplot as plt
from financepy.finutils import *
from financepy.market.curves import *
from financepy.products.equity import *

####################################################################
# FINANCEPY BETA Version 0.180 - This build:  23 Sep 2020 at 01:20 #
#      This software is distributed FREE & WITHOUT ANY WARRANTY    #
# For info and disclaimer - https://github.com/domokane/FinancePy  #
#    For bug reports and comments - allmypythonprojects@gmail.com  #
####################################################################



### Market Data

In [2]:
stockPrice = 100
valueDate = FinDate(1, 9, 2020)
expiryDate = valueDate.addYears(1.0)
dividendYield = 0.00
interestRate = 0.00

K1 = 100
K2 = 105
vol_K1 = 0.20
vol_K2 = 0.20

discountCurve = FinDiscountCurveFlat(valueDate, interestRate, FinFrequencyTypes.CONTINUOUS)

callOptionK1 = FinEquityVanillaOption(expiryDate, K1, FinOptionTypes.EUROPEAN_CALL)
callOptionK2 = FinEquityVanillaOption(expiryDate, K2, FinOptionTypes.EUROPEAN_CALL)

model_vol_K1 = FinEquityModelBlackScholes(vol_K1)
model_vol_K2 = FinEquityModelBlackScholes(vol_K2)    

price_K1 = callOptionK1.value(valueDate, stockPrice, discountCurve, dividendYield, model_vol_K1)
price_K2 = callOptionK2.value(valueDate, stockPrice, discountCurve, dividendYield, model_vol_K2)

call_spread = price_K1 - price_K2

print(call_spread)

2.0599739838503126


### Find the max vol that we can have at K2 when vol at K1 is 20%

In [3]:
max_vol_K2 = callOptionK2.impliedVolatility(valueDate, stockPrice, discountCurve, dividendYield, price_K1)
print(max_vol_K2)

0.25192770245866236


### Check if the value of the call spread is equal to 0 using vol(K2) = 0.25192770245866236

In [4]:
model_max_vol_K2 = FinEquityModelBlackScholes(max_vol_K2) 

price_K2_max_vol = callOptionK2.value(valueDate, stockPrice, discountCurve, dividendYield, model_max_vol_K2)

call_spread_price_smile = price_K1 - price_K2_max_vol
print(call_spread_price_smile)

0.0
